### PR TITLE
fetchpatch: add "decoder" argument for gerrit

### PR DESCRIPTION
**Pull request by alyssais**
###### Motivation for this change

(and gitiles)

This allows fetching a patch from servers that return them
base64-encoded, like this:

    fetchpatch {
      name = "gcc.patch";
      url = "https://chromium.googlesource.com/aosp/platform/external/libchrome/+/f37ae3b1a873d74182a2ac31d96742ead9c1f523^!?format=TEXT";
      decode = "base64 -d";
      sha256 = "11j1bqz2p8xrfzgfrylgdvmqs45489c4ckl7l0ra1dpfgbqy94a8";
    }

This will mean fewer patches have to be directly included in Nixpkgs, permanently bloating the repository size.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/fetchpatch/default.nix [hunk 1]
```diff
@@ -22,6 +31,16 @@ fetchurl ({
       echo "error: Fetched patch file '$out' is empty!" 1>&2
       exit 1
     fi
+
+    ${decode} < "$out" > "$tmpfile"
+    if [ ! -s "$tmpfile" ]; then
+        echo "Failed to decode patch with command \"${decode}\"" >&2
```
[the remark this answers is not in the document]
Comment (alyssais) [reply pkgs/build-support/fetchpatch/default.nix]: Done.  (I changed the shell string to be single quotes to avoid escaping.)

**Comment by alyssais** [comment]
> needs doc in https://nixos.org/manual/nixpkgs/stable/#sec-patches `doc/contributing/coding-conventions.chapter.md`

Done.

**Comment by alyssais** [comment]
Updated to avoid dumping potentially binary files to the terminal if the decoder fails, based on @samueldr's feedback on Matrix, and also made the error text be displayed if the decoder returns non-zero.

**Review by samueldr — APPROVED** [review]
Without having ran the changes, reading carefully it looks all okay.

There shouldn't be any nasty surprises with awkward content (e.g. binary).

If the command fails the author should have useful insight.

**Pull request merged**